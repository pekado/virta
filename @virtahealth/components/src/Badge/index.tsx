import * as React from "react";
import { StyleProp, ViewStyle } from "react-native";
import { BaseTheme, primitives } from "@virtahealth/styles";
import { getPlatformFontFamilySuffix } from "@virtahealth/utils";

import { Base } from "../Text";
import styled from "../styled-components";

interface BadgeProps {
  /**
   * The text that will display within the badge
   */
  label?: string;

  /**
   * The color of the badge. You may use
   * shorthand colors (orange, blue, or pink)
   * any supported HTML colors
   * rgb/hex color codes
   */
  color?: "orange" | "blue" | "pink" | string;

  /**
   * The size used to determine badge size and text size
   */
  size?: number;

  /**
   * Optional styling for the badge container.
   * Primarily intended for positioning the badge
   */
  containerStyle?: StyleProp<ViewStyle>;
}

type BadgeWrapperPassthroughProps = Pick<BadgeProps, "size" | "color">;
type BadgeLabelPassthroughProps = Required<Pick<BadgeProps, "size">>;

const BadgeWrapper = styled.View<BadgeWrapperPassthroughProps>`
  width: ${({ size }) => size}px;
  height: ${({ size }) => size}px;
  background: ${({ color, theme }) => {
    switch (color) {
      case "orange":
        return primitives.color.phosphorousOrange900;
      case "blue":
        return primitives.color.calciumTeal900;
      case "pink":
        return primitives.color.magnesiumPink700;
      default:
        return color || theme.badgeDefaultBackgroundColor;
    }
  }};
  border-radius: ${({ size }) => size}px;
  align-items: center;
  justify-content: center;
`;

const BadgeLabel = styled(Base)<BadgeLabelPassthroughProps>`
  width: 100%;
  color: ${({ theme }) => theme.badgeDefaultTextColor};
  font-weight: ${({ theme }) => theme.badgeBaseTextFontWeight};
  font-family: "${({ theme }) =>
    theme[
      `badgeBaseTextFontFamily${getPlatformFontFamilySuffix()}` as keyof BaseTheme
    ]}";
  text-align: center;
  text-transform: uppercase;
  font-size: ${({ size }) => `${(size * 2) / 3}px`};
`;

export const Badge: React.FC<BadgeProps> = ({
  size = 24,
  label,
  color,
  containerStyle,
}) => (
  <BadgeWrapper size={size} color={color} style={containerStyle}>
    {label !== undefined ? <BadgeLabel size={size}>{label}</BadgeLabel> : null}
  </BadgeWrapper>
);
