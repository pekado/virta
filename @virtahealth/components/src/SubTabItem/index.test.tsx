import * as React from "react";
import { fireEvent } from "@testing-library/react-native";
import { createComponentWithVirtaContext } from "../test";
import { Badge } from "../Badge";
import { SubTabItem } from "./index";

const defaultProps = {
  isSelected: false,
  onPress: jest.fn(),
  testID: "1",
};

describe("Sub Tab Item Component", () => {
  it("Checks that the component renders as unselected", () => {
    const wrapper = createComponentWithVirtaContext(
      <SubTabItem {...defaultProps}>Test Tab</SubTabItem>
    );
    expect(wrapper.getByText("Test Tab")).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it("Checks that the component renders as selected", () => {
    const wrapper = createComponentWithVirtaContext(
      <SubTabItem {...defaultProps} isSelected={true}>
        Test Tab
      </SubTabItem>
    );
    expect(wrapper.getByText("Test Tab")).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it("Triggers the onPress handler", () => {
    const wrapper = createComponentWithVirtaContext(
      <SubTabItem {...defaultProps}>Test Tab</SubTabItem>
    );

    const touchableButton = wrapper.getByTestId("1");

    fireEvent(touchableButton, "press");

    expect(defaultProps.onPress).toBeCalled();
  });

  it("Renders an icon", () => {
    const wrapper = createComponentWithVirtaContext(
      <SubTabItem {...defaultProps} iconAfter={<Badge size={25} label="5" />}>
        Test Tab
      </SubTabItem>
    );
    expect(wrapper.getByText("5")).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });
});
