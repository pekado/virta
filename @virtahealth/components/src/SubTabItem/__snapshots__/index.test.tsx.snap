// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sub Tab Item Component Checks that the component renders as selected 1`] = `
<TouchableOpacity
  isSelected={true}
  onPress={[MockFunction]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderBottomColor": "rgb(34, 110, 173)",
        "borderBottomWidth": 2,
        "borderStyle": "solid",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
  testID="1"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
        },
      ]
    }
  >
    <Text
      isSelected={true}
      style={
        Array [
          Object {
            "color": "rgb(36, 38, 43)",
            "fontFamily": "Whitney-Medium",
            "fontSize": 20,
            "fontWeight": "500",
            "letterSpacing": 0,
            "lineHeight": 20,
          },
          Object {
            "color": "rgb(34, 110, 173)",
            "fontWeight": "700",
            "paddingBottom": 10,
          },
        ]
      }
    >
      Test Tab
    </Text>
  </View>
</TouchableOpacity>
`;

exports[`Sub Tab Item Component Checks that the component renders as unselected 1`] = `
<TouchableOpacity
  isSelected={false}
  onPress={[MockFunction]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderBottomColor": "rgb(168, 174, 182)",
        "borderBottomWidth": 2,
        "borderStyle": "solid",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
  testID="1"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
        },
      ]
    }
  >
    <Text
      isSelected={false}
      style={
        Array [
          Object {
            "color": "rgb(36, 38, 43)",
            "fontFamily": "Whitney-Medium",
            "fontSize": 20,
            "fontWeight": "500",
            "letterSpacing": 0,
            "lineHeight": 20,
          },
          Object {
            "color": "rgb(168, 174, 182)",
            "fontWeight": "700",
            "paddingBottom": 10,
          },
        ]
      }
    >
      Test Tab
    </Text>
  </View>
</TouchableOpacity>
`;

exports[`Sub Tab Item Component Renders an icon 1`] = `
<TouchableOpacity
  isSelected={false}
  onPress={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderBottomColor": "rgb(168, 174, 182)",
        "borderBottomWidth": 2,
        "borderStyle": "solid",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
  testID="1"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
        },
      ]
    }
  >
    <Text
      isSelected={false}
      style={
        Array [
          Object {
            "color": "rgb(36, 38, 43)",
            "fontFamily": "Whitney-Medium",
            "fontSize": 20,
            "fontWeight": "500",
            "letterSpacing": 0,
            "lineHeight": 20,
          },
          Object {
            "color": "rgb(168, 174, 182)",
            "fontWeight": "700",
            "paddingBottom": 10,
          },
        ]
      }
    >
      Test Tab
    </Text>
    <View
      size={25}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgb(22, 56, 76)",
            "borderRadius": 25,
            "height": 25,
            "justifyContent": "center",
            "width": 25,
          },
        ]
      }
    >
      <Text
        size={25}
        style={
          Array [
            Object {
              "color": "rgb(36, 38, 43)",
              "fontFamily": "Whitney-Book",
              "fontSize": 20,
              "fontWeight": "500",
              "letterSpacing": 0,
              "lineHeight": 20,
              "textDecorationColor": "black",
              "textDecorationLine": "none",
              "textDecorationStyle": "solid",
            },
            Object {
              "color": "rgb(255,255,255)",
              "fontFamily": "Whitney-Medium",
              "fontSize": 16.666666666666668,
              "fontWeight": "700",
              "textAlign": "center",
              "textTransform": "uppercase",
              "width": "100%",
            },
          ]
        }
      >
        5
      </Text>
    </View>
  </View>
</TouchableOpacity>
`;
